File: IPL 2021 - Match Day 2.java
class Solution {
    static ArrayList<Integer> max_of_subarrays(int arr[], int n, int k) {
        // Your code here
        ArrayList<Integer>ans=new ArrayList<>();
        Deque<Integer>q=new LinkedList<>();
        for(int i=0;i<n;i++){
            while(q.size()>0 && i-q.peek()>=k){
                q.removeFirst();
            }
            
            while(q.size()>0 && arr[q.getLast()]<=arr[i]){
                q.removeLast();
            }
            q.addLast(i);
            if(i>=k-1)ans.add(arr[q.peek()]);
        }
        
        return ans;
    }
}

File: IPL 2021 - Match Day 2.cpp
class Solution {
  public:
    vector<int> max_of_subarrays(vector<int> arr, int n, int k) {
        // your code here
        vector<int> ans;
        deque<int> dq;
        
        for(int i=0; i<k; i++)
        {
            while(!dq.empty() and arr[dq.back()] <= arr[i])
                dq.pop_back();
            
            dq.push_back(i);
        }
        
        ans.push_back(arr[dq.front()]);
        
        for(int i=k; i<n; i++)
        {
            if(dq.front() <= (i-k))
                dq.pop_front();
                
            while(!dq.empty() and arr[dq.back()] <= arr[i])
                dq.pop_back();
                
            dq.push_back(i);
            
            ans.push_back(arr[dq.front()]);
        }
        
        return ans;
    }
};
